kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/AddFile.java

```diff
@@ -55,42 +61,205 @@ public class AddFile {
+    fieldMap.put(FULL_SCHEMA.indexOf("size"), size);
+    fieldMap.put(FULL_SCHEMA.indexOf("modificationTime"), modificationTime);
+    fieldMap.put(FULL_SCHEMA.indexOf("dataChange"), dataChange);
+    tags.ifPresent(tag -> fieldMap.put(FULL_SCHEMA.indexOf("tags"), tag));
+    baseRowId.ifPresent(id -> fieldMap.put(FULL_SCHEMA.indexOf("baseRowId"), id));
+    defaultRowCommitVersion.ifPresent(
+        version -> fieldMap.put(FULL_SCHEMA.indexOf("defaultRowCommitVersion"), version));
+    stats.ifPresent(stat -> fieldMap.put(FULL_SCHEMA.indexOf("stats"), stat.serializeAsJson()));
+
+    return new GenericRow(FULL_SCHEMA, fieldMap);
+  }
+
+  /** Constructs an {@link AddFile} action from the given 'AddFile' {@link Row}. */
+  public AddFile(Row row) {
+    super(row);
+  }
+
+  @Override
+  protected StructType getFullSchema() {
+    return FULL_SCHEMA;
+  }
+
+  public String getPath() {
+    return (String) getValueAsObject("path");
+  }
+
+  public MapValue getPartitionValues() {
+    return (MapValue) getValueAsObject("partitionValues");
+  }
+
+  public long getSize() {
+    return (long) getValueAsObject("size");
+  }
+
+  public long getModificationTime() {
+    return (long) getValueAsObject("modificationTime");
+  }
+
+  public boolean getDataChange() {
+    return (boolean) getValueAsObject("dataChange");
+  }
+
+  public Optional<DeletionVectorDescriptor> getDeletionVector() {
+    return Optional.ofNullable((Row) getValueAsObject("deletionVector"))
+        .map(DeletionVectorDescriptor::fromRow);
+  }
+
+  public Optional<MapValue> getTags() {
+    return Optional.ofNullable((MapValue) getValueAsObject("tags"));
+  }
+
+  public Optional<Long> getBaseRowId() {
+    return Optional.ofNullable((Long) getValueAsObject("baseRowId"));
+  }
+
+  public Optional<Long> getDefaultRowCommitVersion() {
+    return Optional.ofNullable((Long) getValueAsObject("defaultRowCommitVersion"));
+  }
+
+  public Optional<DataFileStatistics> getStats() {
+    return Optional.ofNullable((String) getValueAsObject("stats"))
+        .flatMap(DataFileStatistics::deserializeFromJson);
+  }
+
+  public Optional<Long> getNumRecords() {
+    return getStats().map(DataFileStatistics::getNumRecords);
+  }
+
+  /** Returns a new {@link AddFile} with the provided baseRowId. */
+  public AddFile withNewBaseRowId(long baseRowId) {
+    return new AddFile(createRowWithOverriddenValue("baseRowId", baseRowId));
+  }
+
+  /** Returns a new {@link AddFile} with the provided defaultRowCommitVersion. */
+  public AddFile withNewDefaultRowCommitVersion(long defaultRowCommitVersion) {
+    return new AddFile(
+        createRowWithOverriddenValue("defaultRowCommitVersion", defaultRowCommitVersion));
+  }
+
+  @Override
+  public String toString() {
+    // Convert the partitionValues and tags to Java Maps and make them sorted by key.
+    Map<String, String> partitionValuesJavaMap = toJavaMap(getPartitionValues());
```

Comment:
This doesn't seem right to me. MapValue doesn't guarantee order, as far as I can tell. Turning this into a Java Map and then a TreeMap certainly also doesn't guarantee they are in the right order.

if the partition columns are, in order, `zzzPart1`, `yyyPart2` ... then tree map will sort them in their natural order, which will be yyyPart2, zzzPart1, which is wrong.

you need to lookup the order as defined in the schema.